### DISCUSSION TITLE
Passing parameter ADDLOCAL from Bundle to MSI doesn't work with WiX 4

### DISCUSSION THREAD

**Discussion by wolfkor** [wixtoolset/issues#7911] [Questions]
### WiX Version

4.0.3

### .NET or MSBuild or Visual Studio Version

.NET 7  - MSBuild-Version 17.8.3+195e7f5a3 für .NET Framework

### HeatWave Version

1.0.2.1

### Windows Version

Win10 22H2

### Repro Steps

We are currently migrating from WiX 3 to WiX 4. In the MSI installer we use ADDLOCAL for selection of feature sets and also allow this by command line call. There is also a bundle installer for installing the prerequisites for the MSI installer. This also processes ADDLOCAL when the command line is called and passes the value to the MSI installer.
`<Variable Name="ADDLOCAL" Value="" bal:Overridable="yes" Type="formatted" />`
...
`<MsiProperty Name="ADDLOCAL" Value="[ADDLOCAL]" />`

### Actual Result

Using WiX 3 this works fine but in WiX 4 it causes the compiler error
`error WIX0365: The 'ADDLOCAL' MsiProperty is controlled by the bootstrapper and cannot be authored. (Illegal properties are: 'ACTION', 'ADDLOCAL', 'ADDSOURCE', 'ADDDEFAULT', 'ADVERTISE', 'ALLUSERS', 'REBOOT', 'REINSTALL', 'REINSTALLMODE', or 'REMOVE'.) Remove the MsiProperty element.`

### Expected Result

Passing parameter ADDLOCAL works

**Comment by barnson** [comment]
WiX doesn't use issues for support. We moved this issue to the [GitHub discussions area][gh]. You might also consider asking your question on [Stack Overflow][so].

[gh]: https://github.com/wixtoolset/issues/discussions?discussions_q=category%3AQuestions
[so]: https://stackoverflow.com/questions/tagged/wix

**Reply by wolfkor** [reply]
@barnson I'm a little bit confused about this. For me this is more of a bug than a support request. Since I believe you are responsible for the change, it would be nice if you not only convert the entry, but also respond to it constructively. If there are good reasons for the change, then you should at least give me the right path to solve my problem. I found the entrys https://stackoverflow.com/questions/43917865/how-to-pass-addlocal-parameter-from-wix-standard-bootstrapper-to-a-msipackage and https://github.com/wixtoolset/issues/issues/5294 with no confirmed solution given by the wix team. At the moment I don't understand why passing ADDLOCAL is a problem.

**Reply by robmen** [reply]
IIRC, there was an issue closed in WiX v4 that explained why those standard Windows Installer properties are not allowed in `MsiProperty`. I expect that will point at the code change as well.

**Comment by demienYu** [comment]
I have a problem pretty much the same as topic starter's. I spent some time googling and stackOverflowing this issue but still have not found a clear answer. To simplify my question I'd rather give you my case as an example.
I want my user to install mongodb before the main package. So I have Bundle.wxs that looks roughly like this:

```
<Bundle Name="My Dojo" Version="0.1.0" UpgradeCode="my-upgrade-code-just-doesnt-matter" Compressed="no">
  <BootstrapperApplication>
      <bal:WixStandardBootstrapperApplication
            LicenseUrl=""
            Theme="hyperlinkLicense" />
  </BootstrapperApplication>
  <Chain>
      <MsiPackage Id="MongoDBPackage"
        SourceFile="mongodb\mongodb-windows-x86_64-7.0.5-signed.msi"
        Description="MongoDB 7.0 Community Edition Database" DisplayName="MongoDB" InstallSize="1000000000" Permanent="yes"
        InstallCondition="ConditionIfMongoDBShouldInstall">
         <MsiProperty Name="SHOULD_INSTALL_COMPASS" Value="0"/>
         <!-- <MsiProperty Name="ADDLOCAL" Value="ServerService"/> -->
      </MsiPackage>
      <MsiPackage Id="MyDojoPackage" Description="My Dojo package description" DisplayName="My Dojo" InstallSize="150000000"
        SourceFile="sandbox-out\my-dojo-package.msi"/>
  </Chain>
</Bundle>
```

MongoDB installation package has a feature called "ServerService". I should set it to "1" if I want MongoDB to be installed as a service.
If I install mongodb separately using msiexec everything works fine.
But I can't install it with required behaviour using wix4.
If I uncomment line <MsiProperty Name="ADDLOCAL" Value="ServerService"/> I will get an error referencing disabled values for MSiProperty (ADDLOCAL in that list of course).

I am realy looking forward "Depl. Dojo season 2" but now I just can not find any solution to my problem.

**Comment by chrpai** [comment]
MSI doesn't allow you to set a feature to "1" from the command line.  You can only set secure custom public properties from the command line.  An examination of the MSI using ORCA reveals the following such properties of interest.

MONGO_DATA_PATH
MONGO_LOG_PATH
MONGO_SERVICE_ACCOUNT_NAME
MONGO_SERVICE_ACCOUNT_PASSWORD
MONGO_SERVICE_ACCOUNT_TYPE
MONGO_SERVICE_INSTALL
MONGO_SERVICE_NAME
SHOULD_INSTALL_COMPASS
WINDOWSBUILDNUM (probably shouldn't be a public property but it is)

I see the MongoDB MSI is written in WiX.  They might want to come here for support if they are boxing themselves into a corner.  Or you could author a transform to beat it into submission.

**Reply by demienYu** [reply]
I think ADDLOCAL really is a secure custom public property. I can install mongodb as a service using command:
msiexec /i mongodb-windows-x86_64-7.0.5-signed.msi MONGO_SERVICE_INSTALL="1" ADDLOCAL="ServerService" SHOULD_INSTALL_COMPASS="0"
In fact "ServerService" is a feature (I inspected msi file with orca).
But the question is: how can I install mongodb as a part of my bundle?
You mentioned that I could author a transform. So let's say I have done it and I have mongodb-7.0.5-modified.mst file.
How can I include it to my bundle?

**Reply by robmen** [reply]
`ADDLOCAL` is a [built-in Windows Installer property](https://learn.microsoft.com/en-us/windows/win32/msi/addlocal) that implies a significant amount of behavior. I'm pretty sure somewhere in [Season 1, Episode 49 of the Deployment Dojo](https://robmensching.com/deployment-dojo/episodes/s1/e49/introducing-features-to-the-dojo-using-wix-v4/) we start exploring the details of it. The discussion might also spill into Episodes 50 and 51. It did take a while to explore all the edge cases.

Ultimately, setting `ADDLOCAL` directly in your bundle exposed bugs in WiX v3 that were fixed in WiX v4 to no longer allow it on the command line. You can control feature states in your custom BA, and Burn will take care of setting the appropriate feature state properties for you during the installation of the MSI.

**Comment by chrpai** [comment]
@robmen  does wix v4 block the INSTALLLEVEL property?
